### PR TITLE
Broadcast queue

### PR DESCRIPTION
**Pull request by girikuncoro**
Merged #44

### CODE REVIEW

File: app/api/sms/broadcast.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    ensure_param
+)
+
+app = Flask(__name__)
+enable_json_error(app)
+
+
+# test credentials
+ACCOUNT_SID = 'AC94da6f9c22530311d9fa507e236832b8'  # add in config?
+AUTH_TOKEN = '359d1bd6ae49c09c3639dde72a79e072'  # add in config?
```
Comment (Ragil) [inline app/api/sms/broadcast.py]: You should never commit credentials into version control. It is now forever stored in history. I recommend that you generate a new auth token from your twilio developer console immediately.

Yes, you should move these values into config and update it manually through the config api.

File: app/api/sms/broadcast_action.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+        phone_numbers = [farmer.phone_number for farmer in farmers]
+
+        if phone_numbers:
+            for phone_number in phone_numbers:
+                taskqueue.add(
+                    queue_name=self.QUEUE_NAME,
+                    url=self.QUEUE_URL,
+                    payload=json.dumps({'phone_number': phone_number,
+                                        'message': cmd.msg})
```
Comment (Ragil) [inline app/api/sms/broadcast_action.py]: The purpose of using a queue is to move long running operations into a background process. During a user request, the user expects a quick turn around on response time otherwise they may think something is not working. In this situation, sending 100-1000 messages to each farmer would be the bottleneck of this user request. If we can move that out of the user request path by telling the queue worker once to send 1000 messages, then we can save a lot of time.

The current implementation tells the queue worker to send 1 message 1000 times, which is not much better than telling twilio to send 1 message 1000 times. Instead we should tell the queue worker to send 1000 messages 1 time. 
How the queue worker sends 1000 messages doesn't matter because it is not running in a user request. In fact, it can simply tell twilio to send 1 message 1000 times (as long as that doesn't take 10 minutes - the default timeout limit for a queue worker)

File: tests/api/sms/test_broadcast_command.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        res_msg = BroadcastAction(cmd).execute()
+
+        self.assertEqual('Pesan berhasil dikirim ke lompoko', _(res_msg))
+
+        tasks = self.taskqueue_stub.get_filtered_tasks()
+        self.assertEqual(len(tasks), 1)  # 1 lompoko farmer in db
+
+        for task in tasks:
+            data = json.loads(task.payload)
+            res = self.app.post(task.url, data=data)  # execute task
```
Comment (Ragil) [inline tests/api/sms/test_broadcast_command.py]: This test file is testing far too much. Test are meant to be very simple and easy to understand. Once you have complexity in the test, they become more of a burden to maintain than the benefit they provide.

We have some good examples of unit test:

- SmsTest validates handling of input and output in the format that twilio expects
- DispatcherTest validates that the dispatcher will dispatch to the appropriate commands

We also have some examples of tests that try and do too much:

- HarvestCommandTest tries to replicate the behaviour of the dispatcher in its test. The same happens in PlantCommandTest 
https://github.com/Ragil/Communicate-Indonesia/blob/master/tests/api/sms/test_harvest_command.py#L103
https://github.com/Ragil/Communicate-Indonesia/blob/master/tests/api/sms/test_plant_command.py#L103

- QueryCommandTest tests both command and action. Whether the QueryCommand can parse input properly is independent of whether the QueryAction can aggregate information correctly. There is also a lot of cleaning up of unnecessary mocks and instantiation of the app.
https://github.com/Ragil/Communicate-Indonesia/blob/master/tests/api/sms/test_query_command.py#L17
https://github.com/Ragil/Communicate-Indonesia/blob/master/tests/api/sms/test_query_command.py#L186

This test file is trying to perform unit and integration tests at the same time. This test case itself, is doing far too much and even trying to replicate the behaviour of the task queue by dispatching the task to the broadcast app.

Test should be very simple. It should define an input and an expected output. When you start having logic and managing state, the test should be broken into smaller tests.

File: app/api/sms/broadcast.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+def broadcast_sms():
+    phone_number = request.form.get('phone_number')
+    message = request.form.get('message')
+
+    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
+
+    msg = client.messages.create(
+        body=message,
+        to=phone_number,
+        from_='+15005550006'  # add in config?
```
Comment (Ragil) [inline app/api/sms/broadcast.py]: This should also be in configuration.

The tests are slow because the tests are actually making an http request to twilio. This should not be the case. The twilio client should be mocked (just like everything appengine) when during the test.

File: tests/api/sms/test_broadcast_command.py (hunk 3 above)
Comment (Ragil) [inline tests/api/sms/test_broadcast_command.py]: Here is a refactor for some of the command and action tests. https://github.com/Ragil/Communicate-Indonesia/pull/48

They should now do two things:
1. test each unit of functionality independently
2. test all the edge cases of each unit of functionality

Integration tests can be very complex and should be kept to a minimum.

I've left the implementation for SellActionTest for you to fill out. As a by product, the SellAction should also be fixed to behave accordingly.

File: app/api/sms/broadcast.py (hunk 1 above)
Comment (girikuncoro) [inline app/api/sms/broadcast.py]: image
Uh noted, I generated new auth token (request secondary token for AuthToken) but I don't think you can do the same for TestAuthToken.

Comment (Ragil) [inline app/api/sms/broadcast.py]: test token does not charge your account.

Comment (girikuncoro) [inline app/api/sms/broadcast.py]: The commit is actually Test AccountSID and Test AuthToken, not the live one.

Comment (Ragil) [inline app/api/sms/broadcast.py]: Its still a bad idea because someone can make a lot of requests using your test token which will result in your account being disabled. I'll generate an new test token for you. Are you using this anywhere else?

Comment (Ragil) [inline app/api/sms/broadcast.py]: You should have a new randomly generated token.

Comment (girikuncoro) [inline app/api/sms/broadcast.py]: Cool thanks, I got new Test AuthToken now. I only use it for this commit, not elsewhere

**Comment by girikuncoro** [comment]
All great explanations!

File: app/api/sms/broadcast_action.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                self.msg = self.args[1]
+
+            if self.args[1] and district:
+                self.to_id = district[0].key.id()
+                self.to = self.args[0]
+                self.msg = self.args[1]
+
+        if sms.user.role == User.ROLE_FARMER:
+            self.to_id = sms.user.district_id
+            self.to = District.get_by_id(self.to_id).name
```
Comment (Ragil) [inline app/api/sms/broadcast_action.py]: to is very confusing. It doesn't describe the type or possible values that the variable can take. Be very explicit naming.

File: app/api/sms/broadcast_action.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    ]
+
+    def __init__(self, sms):
+        super(BroadcastCommand, self).__init__(sms)
+        self.to = None
+        self.msg = None
+        district = []
+
+        if self.args[0]:
+            district = District.query(District.name == self.args[0]).fetch()
```
Comment (Ragil) [inline app/api/sms/broadcast_action.py]: Use district slug to query.
